packages/plugin-client-drizzle/src/http/session.ts

```diff
@@ -0,0 +1,157 @@
+import { SQLPluginResult, SQLQueryResult } from '@xata.io/client';
+import {
+  Logger,
+  NoopLogger,
+  Query,
+  RelationalSchemaConfig,
+  SelectedFieldsOrdered,
+  TablesRelationalConfig,
+  entityKind,
+  fillPlaceholders
+} from 'drizzle-orm';
+import {
+  PgColumn,
+  PgDialect,
+  PgSession,
+  PgTransaction,
+  PgTransactionConfig,
+  PgPreparedQuery,
+  PreparedQueryConfig,
+  QueryResultHKT
+} from 'drizzle-orm/pg-core';
+import { mapResultRow } from '../shared/utils';
+
+export type XataHttpClient = {
+  sql: SQLPluginResult;
+};
+
+export interface QueryResults<ArrayMode extends 'json' | 'array'> {
+  rowCount: number;
+  rows: ArrayMode extends 'array' ? any[][] : Record<string, any>[];
+  rowAsArray: ArrayMode extends 'array' ? true : false;
+}
+
+export class XataHttpPreparedQuery<T extends PreparedQueryConfig> extends PgPreparedQuery<T> {
+  static readonly [entityKind]: string = 'XataHttpPreparedQuery';
+
+  constructor(
+    private client: XataHttpClient,
+    query: Query,
+    private logger: Logger,
+    private fields: SelectedFieldsOrdered<PgColumn> | undefined,
+    private customResultMapper?: (rows: unknown[][]) => T['execute']
+  ) {
+    super(query);
+  }
+
+  async execute(placeholderValues: Record<string, unknown> | undefined = {}): Promise<T['execute']> {
+    const params = fillPlaceholders(this.query.params, placeholderValues);
+
+    this.logger.logQuery(this.query.sql, params);
+
+    const { fields, client, query, customResultMapper } = this;
+
+    if (!fields && !customResultMapper) {
+      const result = await this.client.sql<Record<string, any>>({ statement: query.sql, params });
+      return { rowCount: result.records.length, rows: result.records, rowAsArray: false };
+    }
+
+    const { rows, warning } = await client.sql({ statement: query.sql, params, responseType: 'array' });
+    if (warning) console.warn(warning);
+
+    // @ts-expect-error joinsNotNullableMap is internal
+    const joinsNotNullableMap = this.joinsNotNullableMap;
+
+    return customResultMapper
+      ? customResultMapper(rows as unknown[][])
+      : rows.map((row) => mapResultRow<T['execute']>(fields!, row as unknown[], joinsNotNullableMap));
```

Comment:
Why is the `!` necessary? Should `fields` be added to the conditional, i.e. `customResultMapper && fields`?

Reply:
fields are passed from the constructor, even though they should be defined, if they weren't we should fail. In our codebase we would normally avoid the exclamation mark, but since this is how other adapters handle it in drizzle codebase I'd stick to it. Ideally this code will be removed soon-ish from Xata repo.